torchci/scripts/check_alerts.py

```diff
@@ -0,0 +1,296 @@
+        headers=headers,
+    )
+    data = json.loads(r.text)
+    return data["data"]["repository"]["issues"]["nodes"]
+
+
+def generate_failed_job_issue(failed_jobs: List[JobStatus]) -> Any:
+    issue = {}
+    issue[
+        "title"
+    ] = f"[Pytorch] There are {len(failed_jobs)} Recurrently Failing Jobs on pytorch/pytorch master"
+    body = "Within the last 50 commits, there are the following failures on the master branch of pytorch: \n"
+    for job in failed_jobs:
+        failing_sha = job.failure_chain[-1]["sha"]
+        hud_link = "https://hud.pytorch.org/minihud?name_filter=" + urllib.parse.quote(
+            job.job_name
+        )
+        body += f"- [{job.job_name}]({hud_link}) failed {len(job.failure_chain)} times consecutively starting with "
+        body += f"commit [{failing_sha}](https://hud.pytorch.org/commit/{REPO_OWNER}/{REPO_OWNER}/{failing_sha})"
+        body += "\n\n"
+
+    body += "Please review the errors and revert if needed."
+    issue["body"] = body
+    issue["labels"] = labels
+    issue["assignees"] = ["zengk95"]
+
+    print("Creating alerts for: ", failed_jobs)
+    return issue
+
+
+def create_issue(issue: Any) -> None:
+    requests.post(CREATE_ISSUE_URL, json=issue, headers=headers)
+
+
+def fetch_hud_data() -> Any:
+    response = requests.get(HUD_API_URL)
+    hud_data = json.loads(response.text)
+    return (hud_data["jobNames"], hud_data["shaGrid"])
+
+
+# TODO: Do something about these flaky jobs, save them in rockset or something
+def record_flaky_jobs(flaky_jobs: List[Any]) -> None:
+    return
+
+
+# Creates a Dict of Job Name -> [JobData]. Essentially a Column in HUD
+def map_job_data(jobNames: Any, shaGrid: Any) -> Dict[str, Any]:
+    jobData = {}
+    for sha in shaGrid:
+        for ind, job in enumerate(sha["jobs"]):
+            if jobNames[ind] not in jobData:
+                jobData[jobNames[ind]] = []
+            jobData[jobNames[ind]].append(job)
+    return jobData
+
+
+def is_job_not_pending_or_skipped(job: Any) -> bool:
+    conclusion = job["conclusion"] if "conclusion" in job else None
+    return not (
+        conclusion is None
+        or conclusion == PENDING
+        or conclusion == NEUTRAL
+        or conclusion == SKIPPED
+    )
+
+
+def get_failed_jobs(job_data: List[Any]) -> List[Any]:
+    return [job for job in job_data if job["conclusion"] == "failure"]
+
+
+def categorize_shas(sha_grid: Any) -> List[Tuple[Any, str]]:
+    categorized_shas = []
+    for sha in sha_grid:
+        conclusions = defaultdict(lambda: 0)
+        for job in sha["jobs"]:
+            if "conclusion" in job:
+                conclusions[job["conclusion"]] += 1
+            else:
+                conclusions[SKIPPED] += 1
+        if conclusions[FAILURE] > 0 or conclusions[CANCELED]:
+            categorized_shas.append((sha, FAILURE))
+        elif conclusions[PENDING] > 0:
+            categorized_shas.append((sha, PENDING))
+        elif conclusions[SKIPPED] > 100:
```

Comment:
May be put a comment on why 100 is used here and put it as a constant. I wonder how it can even reach 100 here when the PR body says something like "Within the last 50 commits"

Reply:
So this is categorizing the failures by SHA, so a certain SHA could skip all the jobs if there's another commit that's added right before it. This happens often times in the case of ghstack where they don't run the CI for everything in the stack, but just one of them.